File: goproxy/forward_test.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

type capturedRequest struct {
	path    string
	query   string
	headers http.Header
	body    map[string]interface{}
}

// newProxyUpstream returns a fake upstream plus a pointer to the last request
// it saw, so tests can assert on what the proxy actually forwarded.
func newProxyUpstream(t *testing.T, handle func(w http.ResponseWriter, r *http.Request)) (*httptest.Server, *capturedRequest) {
	t.Helper()
	captured := &capturedRequest{}
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		captured.path = r.URL.Path
		captured.query = r.URL.RawQuery
		captured.headers = r.Header.Clone()
		var body map[string]interface{}
		if err := json.NewDecoder(r.Body).Decode(&body); err == nil {
			captured.body = body
		}
		handle(w, r)
	}))
	t.Cleanup(server.Close)
	return server, captured
}

func proxyConfig(baseURL string) Config {
	return Config{
		TargetBaseURL:         baseURL,
		TargetChatPath:        "/v1/chat/completions",
		TargetMessagesPath:    "/v1/messages",
		TargetEmbeddingsPath:  "/v1/embeddings",
		TargetCompletionsPath: "/v1/completions",
		ProxyModelName:        "claude-opus-5",
		TargetModelName:       "test-target",
		ProviderAPIKey:        "upstream-key",
	}
}

func postJSON(t *testing.T, handler http.Handler, path string, payload map[string]interface{}, headers map[string]string) *httptest.ResponseRecorder {
	t.Helper()
	body, err := json.Marshal(payload)
	if err != nil {
		t.Fatal(err)
	}
	req := httptest.NewRequest(http.MethodPost, path, bytes.NewReader(body))
	req.Header.Set("Content-Type", "application/json")
	for k, v := range headers {
		req.Header.Set(k, v)
	}
	rec := httptest.NewRecorder()
	handler.ServeHTTP(rec, req)
	return rec
}

func TestChatCompletionsMasksModelBothWays(t *testing.T) {
	upstream, captured := newProxyUpstream(t, func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		fmt.Fprint(w, `{"id":"chatcmpl-1","model":"test-target","choices":[{"index":0,"message":{"role":"assistant","content":"hi"},"finish_reason":"stop"}]}`)
	})
	cfg := proxyConfig(upstream.URL)
	handler := forwardToUpstream(cfg, cfg.TargetChatPath)

	rec := postJSON(t, handler, "/v1/chat/completions", map[string]interface{}{
		"model":       "claude-opus-5",
		"messages":    []interface{}{map[string]interface{}{"role": "user", "content": "hello"}},
		"temperature": 0.5,
	}, nil)

	if rec.Code != http.StatusOK {
		t.Fatalf("status = %d, want 200; body: %s", rec.Code, rec.Body.String())
	}
	if captured.body["model"] != "test-target" {
		t.Errorf("upstream model = %v, want test-target", captured.body["model"])
	}
	if captured.body["temperature"] != 0.5 {
		t.Errorf("temperature = %v, want it forwarded unchanged", captured.body["temperature"])
	}
	// An OpenAI-shaped request must reach the upstream untranslated.
	if _, hasSystem := captured.body["system"]; hasSystem {
		t.Error("chat completions request should not be rewritten into Anthropic shape")
	}

	var resp map[string]interface{}
	if err := json.Unmarshal(rec.Body.Bytes(), &resp); err != nil {
		t.Fatalf("response is not valid JSON: %v", err)
	}
	if resp["model"] != "claude-opus-5" {
		t.Errorf("response model = %v, want claude-opus-5", resp["model"])
	}
	if resp["choices"] == nil {
		t.Error("chat completions response should keep its OpenAI shape")
	}
}

func TestForwardMasksAnyAnthropicModelName(t *testing.T) {
	upstream, captured := newProxyUpstream(t, func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		fmt.Fprint(w, `{"model":"test-target"}`)
	})
	cfg := proxyConfig(upstream.URL)
	handler := forwardToUpstream(cfg, cfg.TargetChatPath)

	postJSON(t, handler, "/v1/chat/completions", map[string]interface{}{
		"model":    "claude-sonnet-4-5",
		"messages": []interface{}{},
	}, nil)

	if captured.body["model"] != "test-target" {
		t.Errorf("upstream model = %v, want test-target", captured.body["model"])
	}
}

func TestForwardLeavesForeignModelAlone(t *testing.T) {
	upstream, captured := newProxyUpstream(t, func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		fmt.Fprint(w, `{"model":"gpt-4o"}`)
	})
	cfg := proxyConfig(upstream.URL)
	handler := forwardToUpstream(cfg, cfg.TargetChatPath)

	rec := postJSON(t, handler, "/v1/chat/completions", map[string]interface{}{
		"model":    "gpt-4o",
		"messages": []interface{}{},
	}, nil)

	if captured.body["model"] != "gpt-4o" {
		t.Errorf("upstream model = %v, want gpt-4o untouched", captured.body["model"])
	}
	if !strings.Contains(rec.Body.String(), "gpt-4o") {
		t.Errorf("response = %s, want gpt-4o untouched", rec.Body.String())
	}
}

func TestForwardUsesProviderKeyAndDropsClientCredentials(t *testing.T) {
	upstream, captured := newProxyUpstream(t, func(w http.ResponseWriter, r *http.Request) {
		fmt.Fprint(w, `{}`)
	})
	cfg := proxyConfig(upstream.URL)
	handler := forwardToUpstream(cfg, cfg.TargetChatPath)

	postJSON(t, handler, "/v1/chat/completions", map[string]interface{}{"model": "claude-opus-5"}, map[string]string{
		"Authorization": "Bearer client-proxy-key",
		"X-Api-Key":     "client-proxy-key",
	})

	if got := captured.headers.Get("Authorization"); got != "Bearer upstream-key" {
		t.Errorf("upstream Authorization = %q, want the provider key", got)
	}
	if got := captured.headers.Get("X-Api-Key"); got != "" {
		t.Errorf("upstream X-Api-Key = %q, want the client credential stripped", got)
	}
}

func TestForwardStreamPassesSSEThroughWithMaskedModel(t *testing.T) {
	upstream, captured := newProxyUpstream(t, func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "text/event-stream")
		w.WriteHeader(http.StatusOK)
		fmt.Fprint(w, "data: {\"id\":\"chatcmpl-1\",\"model\":\"test-target\",\"choices\":[{\"delta\":{\"content\":\"hi\"}}]}\n\n")
		fmt.Fprint(w, "data: [DONE]\n\n")
	})
	cfg := proxyConfig(upstream.URL)
	handler := forwardToUpstream(cfg, cfg.TargetChatPath)

	rec := postJSON(t, handler, "/v1/chat/completions", map[string]interface{}{
		"model":  "claude-opus-5",
		"stream": true,
	}, nil)

	if ct := rec.Header().Get("Content-Type"); ct != "text/event-stream" {
		t.Errorf("Content-Type = %q, want text/event-stream", ct)
	}
	if captured.headers.Get("Accept") != "text/event-stream" {
		t.Errorf("upstream Accept = %q, want text/event-stream", captured.headers.Get("Accept"))
	}
	body := rec.Body.String()
	if strings.Contains(body, "test-target") {
		t.Errorf("streamed body leaks the upstream model: %s", body)
	}
	if !strings.Contains(body, `"model":"claude-opus-5"`) {
		t.Errorf("streamed body = %s, want the masked model", body)
	}
	// OpenAI clients on this route must still receive OpenAI chunks.
	if !strings.Contains(body, `"choices"`) || strings.Contains(body, "content_block_delta") {
		t.Errorf("streamed body should stay in OpenAI shape: %s", body)
	}
	if !strings.HasSuffix(strings.TrimRight(body, "\n"), "data: [DONE]") {
		t.Errorf("streamed body should end with [DONE]: %q", body)
	}
}

func TestWildcardForwardsInboundPathAndQuery(t *testing.T) {
	upstream, captured := newProxyUpstream(t, func(w http.ResponseWriter, r *http.Request) {
		fmt.Fprint(w, `{}`)
	})
	cfg := proxyConfig(upstream.URL)
	handler := forwardToUpstream(cfg, "")

	req := httptest.NewRequest(http.MethodGet, "/v1/files?limit=2", nil)
	rec := httptest.NewRecorder()
	handler.ServeHTTP(rec, req)

	if captured.path != "/v1/files" {
		t.Errorf("upstream path = %q, want /v1/files", captured.path)
	}
	if captured.query != "limit=2" {
		t.Errorf("upstream query = %q, want limit=2", captured.query)
	}
}

func TestForwardRejectsInvalidJSON(t *testing.T) {
	upstream, _ := newProxyUpstream(t, func(w http.ResponseWriter, r *http.Request) {
		t.Error("upstream should not be reached for invalid JSON")
	})
	cfg := proxyConfig(upstream.URL)
	handler := forwardToUpstream(cfg, cfg.TargetChatPath)

	req := httptest.NewRequest(http.MethodPost, "/v1/chat/completions", strings.NewReader("not json"))
	req.Header.Set("Content-Type", "application/json")
	rec := httptest.NewRecorder()
	handler.ServeHTTP(rec, req)

	if rec.Code != http.StatusBadRequest {
		t.Errorf("status = %d, want 400", rec.Code)
	}
}

func TestAuthMiddlewareAcceptsAnthropicApiKeyHeader(t *testing.T) {
	cfg := Config{ProxyAPIKey: "secret"}
	handler := authMiddleware(cfg, func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusOK)
	})

	for name, header := range map[string]map[string]string{
		"bearer":    {"Authorization": "Bearer secret"},
		"x-api-key": {"X-Api-Key": "secret"},
	} {
		req := httptest.NewRequest(http.MethodPost, "/v1/messages", nil)
		for k, v := range header {
			req.Header.Set(k, v)
		}
		rec := httptest.NewRecorder()
		handler.ServeHTTP(rec, req)
		if rec.Code != http.StatusOK {
			t.Errorf("%s: status = %d, want 200", name, rec.Code)
		}
	}

	req := httptest.NewRequest(http.MethodPost, "/v1/messages", nil)
	req.Header.Set("Authorization", "Bearer wrong")
	rec := httptest.NewRecorder()
	handler.ServeHTTP(rec, req)
	if rec.Code != http.StatusUnauthorized {
		t.Errorf("status = %d, want 401", rec.Code)
	}
}

File: nodeproxy/forward.test.js
/**
 * End-to-end tests: a real upstream HTTP server behind the real proxy, driven
 * over the network. node-fetch is deliberately NOT mocked here, so streaming,
 * headers and status codes are exercised for real.
 */

const http = require("http");

const originalEnv = process.env;

/** Records the last request and replays a scripted response. */
function startUpstream() {
  const state = { path: null, headers: null, body: null, routes: {} };
  const server = http.createServer((req, res) => {
    const chunks = [];
    req.on("data", (chunk) => chunks.push(chunk));
    req.on("end", () => {
      const raw = Buffer.concat(chunks).toString();
      state.path = req.url;
      state.headers = req.headers;
      try {
        state.body = raw ? JSON.parse(raw) : null;
      } catch (e) {
        state.body = null;
      }

      const route = state.routes[req.url.split("?")[0]];
      if (!route) {
        res.writeHead(404, { "Content-Type": "application/json" });
        res.end(JSON.stringify({ error: { message: "no such route" } }));
        return;
      }
      res.writeHead(route.status, { "Content-Type": route.contentType });
      res.end(route.payload);
    });
  });
  state.server = server;
  state.respond = (path, payload, { status = 200, contentType = "application/json" } = {}) => {
    state.routes[path] = { status, contentType, payload };
  };
  return state;
}

function listen(server) {
  return new Promise((resolve) => server.listen(0, "127.0.0.1", () => resolve(server.address().port)));
}

function close(server) {
  return new Promise((resolve) => (server ? server.close(resolve) : resolve()));
}

function parseSSEEvents(raw) {
  const events = [];
  for (const block of raw.split("\n\n")) {
    const trimmed = block.trim();
    if (!trimmed) continue;
    let eventType = "";
    let data = null;
    for (const line of trimmed.split("\n")) {
      if (line.startsWith("event:")) eventType = line.slice(6).trim();
      else if (line.startsWith("data:")) data = line.slice(5).trim();
    }
    if (!data || data === "[DONE]") continue;
    const parsed = JSON.parse(data);
    parsed.__event = eventType;
    events.push(parsed);
  }
  return events;
}

describe("proxy end to end", () => {
  let upstream;
  let proxy;
  let proxyUrl;

  async function startProxy(env = {}) {
    jest.resetModules();
    process.env = {
      ...originalEnv,
      SERVER_PORT: "0",
      TARGET_BASE_URL: `http://127.0.0.1:${upstream.server.address().port}`,
      PROXY_MODEL_NAME: "claude-opus-5",
      TARGET_MODEL_NAME: "test-target",
      PROVIDER_API_KEY: "upstream-key",
      ...env,
    };
    delete process.env.PROXY_API_KEY;
    const { startServer } = require("./main");
    proxy = startServer();
    await new Promise((resolve) => proxy.on("listening", resolve));
    proxyUrl = `http://127.0.0.1:${proxy.address().port}`;
  }

  function post(path, payload, headers = {}) {
    return fetch(proxyUrl + path, {
      method: "POST",
      headers: { "Content-Type": "application/json", ...headers },
      body: typeof payload === "string" ? payload : JSON.stringify(payload),
    });
  }

  beforeEach(async () => {
    upstream = startUpstream();
    await listen(upstream.server);
  });

  afterEach(async () => {
    await close(proxy);
    await close(upstream.server);
    proxy = null;
    process.env = originalEnv;
  });

  describe("OpenAI routes are forwarded, not translated", () => {
    test("only the model name is rewritten, in both directions", async () => {
      upstream.respond(
        "/v1/chat/completions",
        JSON.stringify({
          id: "chatcmpl-1",
          model: "test-target",
          choices: [{ index: 0, message: { role: "assistant", content: "hi" }, finish_reason: "stop" }],
        }),
      );
      await startProxy();

      const res = await post("/v1/chat/completions", {
        model: "claude-opus-5",
        messages: [{ role: "user", content: "hello" }],
        temperature: 0.5,
      });
      const body = await res.json();

      expect(res.status).toBe(200);
      expect(upstream.path).toBe("/v1/chat/completions");
      expect(upstream.body.model).toBe("test-target");
      expect(upstream.body.temperature).toBe(0.5);
      expect(upstream.body.messages).toEqual([{ role: "user", content: "hello" }]);
      expect(upstream.body.system).toBeUndefined();

      expect(body.model).toBe("claude-opus-5");
      expect(body.choices).toBeDefined();
      expect(body.content).toBeUndefined();
    });

    test("any anthropic model name is masked, foreign ones are left alone", async () => {
      upstream.respond("/v1/chat/completions", JSON.stringify({ model: "test-target" }));
      await startProxy();

      await post("/v1/chat/completions", { model: "claude-sonnet-4-5", messages: [] });
      expect(upstream.body.model).toBe("test-target");

      await post("/v1/chat/completions", { model: "gpt-4o", messages: [] });
      expect(upstream.body.model).toBe("gpt-4o");
    });

    test("streamed chunks stay in OpenAI shape with the model masked", async () => {
      upstream.respond(
        "/v1/chat/completions",
        'data: {"id":"chatcmpl-1","model":"test-target","choices":[{"delta":{"content":"hi"}}]}\n\ndata: [DONE]\n\n',
        { contentType: "text/event-stream" },
      );
      await startProxy();

      const res = await post("/v1/chat/completions", { model: "claude-opus-5", stream: true });
      const raw = await res.text();

      expect(res.headers.get("content-type")).toContain("text/event-stream");
      expect(upstream.headers.accept).toBe("text/event-stream");
      expect(raw).not.toContain("test-target");
      expect(raw).toContain('"model":"claude-opus-5"');
      expect(raw).toContain('"choices"');
      expect(raw).not.toContain("content_block_delta");
      expect(raw.trim().endsWith("data: [DONE]")).toBe(true);
    });

    test("the caller's credential is replaced, never relayed", async () => {
      upstream.respond("/v1/chat/completions", "{}");
      await startProxy();

      await post(
        "/v1/chat/completions",
        { model: "claude-opus-5" },
        { Authorization: "Bearer client-key", "x-api-key": "client-key" },
      );

      expect(upstream.headers.authorization).toBe("Bearer upstream-key");
      expect(upstream.headers["x-api-key"]).toBeUndefined();
    });

    test("malformed JSON is rejected before the upstream is called", async () => {
      upstream.respond("/v1/chat/completions", "{}");
      await startProxy();

      const res = await post("/v1/chat/completions", "not json");

      expect(res.status).toBe(400);
      expect(upstream.path).toBeNull();
    });

    test("each endpoint reaches its configured target path", async () => {
      await startProxy({
        TARGET_CHAT_PATH: "/custom/chat",
        TARGET_RESPONSES_PATH: "/custom/responses",
        TARGET_EMBEDDINGS_PATH: "/custom/embeddings",
        TARGET_MODERATIONS_PATH: "/custom/moderations",
        TARGET_COMPLETIONS_PATH: "/custom/completions",
      });

      const cases = {
        "/v1/chat/completions": "/custom/chat",
        "/v1/responses": "/custom/responses",
        "/v1/embeddings": "/custom/embeddings",
        "/v1/moderations": "/custom/moderations",
        "/v1/completions": "/custom/completions",
      };
      for (const [inbound, target] of Object.entries(cases)) {
        upstream.respond(target, "{}");
        upstream.path = null;
        await post(inbound, { model: "claude-opus-5" });
        expect(upstream.path).toBe(target);
      }
    });

    test("an oversized body is rejected with a JSON 413", async () => {
      upstream.respond("/v1/chat/completions", "{}");
      await startProxy({ MAX_BODY_MB: "1" });

      const res = await post("/v1/chat/completions", {
        model: "claude-opus-5",
        pad: "x".repeat(2 * 1024 * 1024),
      });

      expect(res.status).toBe(413);
      expect(res.headers.get("content-type")).toContain("application/json");
      expect((await res.json()).error.message).toBe("Request body too large");
      expect(upstream.path).toBeNull();
    });

    test("a body under the limit is forwarded", async () => {
      upstream.respond("/v1/chat/completions", "{}");
      await startProxy({ MAX_BODY_MB: "1" });

      await post("/v1/chat/completions", { model: "claude-opus-5", pad: "x".repeat(1024) });

      expect(upstream.path).toBe("/v1/chat/completions");
    });

    test("an unreachable upstream becomes a 502, not a crash", async () => {
      await startProxy({ TARGET_BASE_URL: "http://127.0.0.1:1" });

      const res = await post("/v1/chat/completions", { model: "claude-opus-5" });

      expect(res.status).toBe(502);
      expect((await res.json()).error).toBeDefined();
    });

    test("an upstream error status is preserved", async () => {
      upstream.respond("/v1/chat/completions", JSON.stringify({ error: { message: "nope" } }), {
        status: 400,
      });
      await startProxy();

      const res = await post("/v1/chat/completions", { model: "claude-opus-5" });

      expect(res.status).toBe(400);
      expect(await res.text()).toContain("nope");
    });

    test("the wildcard route keeps the inbound path and query", async () => {
      upstream.respond("/v1/files", JSON.stringify({ object: "list" }));
      await startProxy();

      const res = await fetch(`${proxyUrl}/v1/files?limit=2`);

      expect(res.status).toBe(200);
      expect(upstream.path).toBe("/v1/files?limit=2");
    });
  });

  describe("/v1/models", () => {
    test("is served locally without calling the upstream", async () => {
      await startProxy({
        PROXY_MODEL_ID: "test-id",
        PROXY_MODEL_OWNED_BY: "test-owner",
        PROXY_MODEL_CREATED: "1789100219",
      });

      const res = await fetch(`${proxyUrl}/v1/models`);
      const body = await res.json();

      expect(res.status).toBe(200);
      expect(body.object).toBe("list");
      expect(body.data).toEqual([
        { id: "test-id", object: "model", created: 1789100219, owned_by: "test-owner" },
      ]);
      expect(upstream.path).toBeNull();
    });

    test("a non-GET is forwarded upstream", async () => {
      upstream.respond("/v1/models", JSON.stringify({ object: "list", data: [] }));
      await startProxy();

      const res = await post("/v1/models", { model: "claude-opus-5" });

      expect(res.status).toBe(200);
      expect(upstream.path).toBe("/v1/models");
    });
  });

  describe("/v1/messages relay", () => {
    test("an anthropic model is relayed through chat completions", async () => {
      upstream.respond(
        "/v1/chat/completions",
        JSON.stringify({
          id: "chatcmpl-abc123",
          model: "test-target",
          choices: [
            { index: 0, message: { role: "assistant", content: "Hi from upstream" }, finish_reason: "stop" },
          ],
          usage: { prompt_tokens: 10, completion_tokens: 5 },
        }),
      );
      await startProxy();

      const res = await post("/v1/messages", {
        model: "claude-opus-4-5",
        max_tokens: 100,
        system: "You are helpful.",
        messages: [{ role: "user", content: "Hello" }],
      });
      const body = await res.json();

      expect(upstream.path).toBe("/v1/chat/completions");
      expect(upstream.body.model).toBe("test-target");
      expect(upstream.body.max_tokens).toBe(100);
      expect(upstream.body.messages[0]).toEqual({ role: "system", content: "You are helpful." });

      expect(body.type).toBe("message");
      expect(body.role).toBe("assistant");
      expect(body.model).toBe("claude-opus-5");
      expect(body.id).toBe("msg_abc123");
      expect(body.content).toEqual([{ type: "text", text: "Hi from upstream" }]);
      expect(body.stop_reason).toBe("end_turn");
      expect(body.usage).toEqual({ input_tokens: 10, output_tokens: 5 });
    });

    test("a streamed relay produces anthropic events and closes the response", async () => {
      upstream.respond(
        "/v1/chat/completions",
        [
          'data: {"id":"chatcmpl-s1","choices":[{"index":0,"delta":{"role":"assistant","content":"Hi"}}]}',
          'data: {"id":"chatcmpl-s1","choices":[{"index":0,"delta":{"content":" there"}}]}',
          'data: {"id":"chatcmpl-s1","choices":[{"index":0,"delta":{},"finish_reason":"stop"}],"usage":{"completion_tokens":2}}',
          "data: [DONE]",
        ].join("\n\n") + "\n\n",
        { contentType: "text/event-stream" },
      );
      await startProxy();

      const res = await post("/v1/messages", {
        model: "claude-sonnet-4-5",
        stream: true,
        messages: [{ role: "user", content: "Hello" }],
      });
      // This read only returns because the proxy ends the response.
      const raw = await res.text();

      expect(res.headers.get("content-type")).toContain("text/event-stream");
      expect(upstream.body.stream).toBe(true);

      const events = parseSSEEvents(raw);
      expect(events.map((e) => e.__event)).toEqual([
        "message_start",
        "content_block_start",
        "content_block_delta",
        "content_block_delta",
        "content_block_stop",
        "message_delta",
        "message_stop",
      ]);
      const text = events
        .filter((e) => e.__event === "content_block_delta")
        .map((e) => e.delta.text)
        .join("");
      expect(text).toBe("Hi there");
      expect(events[0].message.model).toBe("claude-opus-5");
    });

    test("a non-anthropic model uses the upstream Messages API", async () => {
      upstream.respond(
        "/v1/messages",
        JSON.stringify({ type: "message", content: [{ type: "text", text: "direct" }] }),
      );
      await startProxy();

      const res = await post("/v1/messages", {
        model: "gpt-4o",
        messages: [{ role: "user", content: "Hello" }],
      });

      expect(upstream.path).toBe("/v1/messages");
      expect(await res.text()).toContain("direct");
    });

    test("a 404 from the upstream Messages API falls back to translation", async () => {
      upstream.respond(
        "/v1/chat/completions",
        JSON.stringify({
          id: "chatcmpl-fallback",
          choices: [
            { index: 0, message: { role: "assistant", content: "translated" }, finish_reason: "stop" },
          ],
        }),
      );
      await startProxy();

      const res = await post("/v1/messages", {
        model: "gpt-4o",
        messages: [{ role: "user", content: "Hello" }],
      });
      const body = await res.json();

      expect(upstream.path).toBe("/v1/chat/completions");
      expect(body.type).toBe("message");
      expect(body.model).toBe("claude-opus-5");
      expect(body.content).toEqual([{ type: "text", text: "translated" }]);
    });

    test("passthrough mode never translates", async () => {
      upstream.respond(
        "/v1/messages",
        JSON.stringify({ type: "message", content: [{ type: "text", text: "direct" }] }),
      );
      await startProxy({ MESSAGES_UPSTREAM_MODE: "passthrough" });

      const res = await post("/v1/messages", {
        model: "claude-opus-4-5",
        messages: [{ role: "user", content: "Hello" }],
      });

      expect(upstream.path).toBe("/v1/messages");
      expect(await res.text()).toContain("direct");
    });

    test("an upstream error comes back in the anthropic envelope", async () => {
      upstream.respond(
        "/v1/chat/completions",
        JSON.stringify({ error: { message: "slow down", type: "rate_limit_exceeded" } }),
        { status: 429 },
      );
      await startProxy();

      const res = await post("/v1/messages", {
        model: "claude-opus-4-5",
        messages: [{ role: "user", content: "Hello" }],
      });
      const body = await res.json();

      expect(res.status).toBe(429);
      expect(body.type).toBe("error");
      expect(body.error.type).toBe("rate_limit_error");
      expect(body.error.message).toBe("slow down");
    });

    test("an unreachable upstream becomes a 502, not a crash", async () => {
      await startProxy({ TARGET_BASE_URL: "http://127.0.0.1:1" });

      const res = await post("/v1/messages", {
        model: "claude-opus-4-5",
        messages: [{ role: "user", content: "Hello" }],
      });
      const body = await res.json();

      expect(res.status).toBe(502);
      expect(body.type).toBe("error");
    });
  });

  describe("proxy authentication", () => {
    test("accepts a bearer token or an x-api-key, rejects anything else", async () => {
      upstream.respond("/v1/chat/completions", "{}");
      await startProxy();
      // startProxy deletes PROXY_API_KEY, so re-start with it set.
      await close(proxy);
      jest.resetModules();
      process.env = {
        ...process.env,
        SERVER_PORT: "0",
        TARGET_BASE_URL: `http://127.0.0.1:${upstream.server.address().port}`,
        PROXY_API_KEY: "secret",
        PROVIDER_API_KEY: "upstream-key",
      };
      const { startServer } = require("./main");
      proxy = startServer();
      await new Promise((resolve) => proxy.on("listening", resolve));
      proxyUrl = `http://127.0.0.1:${proxy.address().port}`;

      expect((await post("/v1/chat/completions", {}, { Authorization: "Bearer secret" })).status).toBe(200);
      expect((await post("/v1/chat/completions", {}, { "x-api-key": "secret" })).status).toBe(200);
      expect((await post("/v1/chat/completions", {}, { Authorization: "Bearer wrong" })).status).toBe(401);
      expect((await post("/v1/chat/completions", {})).status).toBe(401);
    });
  });
});
